src/reactviews/pages/QueryResult/table/plugins/contextMenu.plugin.ts

```diff
@@ -116,24 +118,90 @@ export class ContextMenu<T extends Slick.SlickData> {
+                            batchId: this.resultSetSummary.batchId,
+                            resultId: this.resultSetSummary.id,
+                            selection: selection,
+                            headersFlag: false,
+                        },
+                    );
+                } else {
+                    await this.webViewState.extensionRpc.call("copySelection", {
+                        uri: this.uri,
+                        batchId: this.resultSetSummary.batchId,
+                        resultId: this.resultSetSummary.id,
+                        selection: selection,
+                    });
+                }
 
                 console.log("Copy action triggered");
                 break;
             case "copy-with-headers":
-                await this.webViewState.extensionRpc.call("copyWithHeaders", {
-                    uri: this.uri,
-                    batchId: this.resultSetSummary.batchId,
-                    resultId: this.resultSetSummary.id,
-                    selection: selection,
-                });
+                if (this.dataProvider.isDataInMemory) {
+                    let range = selectionToRange(selection[0]);
+                    let data = await this.dataProvider.getRangeAsync(
+                        range.start,
+                        range.length,
+                    );
+                    const dataArray = data.map((map) => {
+                        const maxKey = Math.max(
+                            ...Array.from(Object.keys(map)).map(Number),
+                        ); // Get the maximum key
+                        return Array.from(
+                            { length: maxKey + 1 },
+                            (_, index) => ({
+                                rowId: index,
+                                displayValue: map[index].displayValue || null,
+                            }),
+                        );
+                    });
+
+                    await this.webViewState.extensionRpc.call(
+                        "sendToClipboard",
+                        {
+                            uri: this.uri,
+                            data: dataArray,
+                            batchId: this.resultSetSummary.batchId,
+                            resultId: this.resultSetSummary.id,
+                            selection: selection,
+                            headersFlag: true,
+                        },
+                    );
+                } else {
+                    await this.webViewState.extensionRpc.call(
+                        "copyWithHeaders",
+                        {
+                            uri: this.uri,
+                            batchId: this.resultSetSummary.batchId,
+                            resultId: this.resultSetSummary.id,
+                            selection: selection,
+                        },
+                    );
+                }
 
-                console.log("Copy with Headers action triggered");
```

Comment:
Is deleting this line intentional? I see you kept the other logs for these events.
 I guess we need some sort of logger at some point....

Reply:
Yeah, I'll clean up these in a separate PR.